### PR TITLE
feat(encoding): add Better compression level (zstd level 7, lazy2 strategy)

### LINKED ISSUE #6: feat: implement Better compression level (zstd level 7, lazy2 strategy)
## Summary

`CompressionLevel::Better` currently triggers `unimplemented!()`. Level 7 uses lazy2 strategy — significantly better compression ratio than dfast at moderate speed cost.

### C reference implementation (zstd level 7)
- **Strategy**: `ZSTD_lazy2` (lazy matching with 2-step look-ahead)
- **Parameters**: `windowLog=23, chainLog=23, hashLog=22, searchLog=3, minMatch=3, targetLength=48`
- **Key files**: `lib/compress/zstd_lazy.c`

### Algorithm
1. **Hash chain table** — hash → linked list of positions with same hash
2. **Lazy matching** — at each position:
   - Find best match at position N
   - Find best match at position N+1 (with 1 extra literal)
   - Find best match at position N+2 (with 2 extra literals)
   - Pick the best overall (longer match justifies extra literals)
3. **Common-length caching** — when traversing hash chains, skip bytes already compared (critical optimization from C code)
4. **Row matching** — `ZSTD_RowMatchState` with SIMD tag matching for faster chain traversal

### What needs to be implemented

1. **Hash chain data structure** — position chains indexed by hash
2. **Lazy2 match finder** — 2-step look-ahead with price comparison
3. **Match length comparison** — efficient byte-level comparison with early exit
4. **Chain depth limiting** — `searchLog=3` means max 8 chain entries checked
5. **Rep code integration** — check repeat offsets at each lazy step

### Acceptance criteria
- [ ] `CompressionLevel::Better` no longer panics
- [ ] Roundtrip + interop tests pass
- [ ] Compression ratio within 10% of C zstd level 7 on silesia corpus
- [ ] Speed slower than Default but ratio measurably better

### Dependencies
- #5 (Default level) — shared infrastructure for hash tables and match counting

## Time estimate
4d

## Blocked by
- #5 (Default level — shared hash table and match counting infrastructure)

### PR DESCRIPTION
**Pull request by polaz**
## Summary

- Implement `CompressionLevel::Better` — roughly equivalent to zstd level 7
- Uses a **hash chain matcher** (hash table + chain table) with lazy2 strategy and gain-based match cost model
- Explores up to 16 match candidates per position (vs 4 in Default's Dfast buckets)
- 8 MB window, 6 MB hash/chain tables, ~22 MB peak memory

## Technical Details

**Hash chain architecture** (`HcMatchGenerator`):
- Hash table (1M u32 entries, 4 MB) maps hash → most recent position
- Chain table (512K u32 entries, 2 MB) links positions with the same hash
- Chain traversal skips stale entries and explores up to 16 valid candidates per position (matching C zstd level 7's `searchLog=4`)
- Gain-based lazy2 matching: `match_len × 4 − log2(offset)` cost model with +4 literal bonus
- Boundary backfill: last 3 positions of each slice are seeded when the next slice arrives

**Compression ratio** (corpus z000033, 1 MB):
- Default (Dfast): 578,142 bytes
- **Better (hash chain): 541,592 bytes** (+6.3% vs Default)
- C zstd level 7: 510,030 bytes (our Better is within 6.2%)

## Known Limitations

- ~6% ratio gap vs C zstd level 7 remains; C uses deeper chain traversal and row-based SIMD matching
- Small/highly-repetitive data shows minimal improvement over Default (already near-optimal)

## Test Plan

- 10 roundtrip integrity tests (compressible, random, multi-block, streaming, edge cases, repeat offsets, large literals, 8 MiB window boundary)
- Compression ratio tests: Better < Default on corpus, Better within 1% of Default on repeat-offset fixture
- Cross-validation: Rust Better compress → C FFI decompress
- Streaming encoder roundtrip + buffer reuse test
- All 182 workspace tests pass

Closes #6

## Summary by CodeRabbit

* **New Features**
  * "Better" compression (CLI level 3) is now supported for one-shot and streaming use, routing through the shared block-encoded pipeline and accepted by the CLI.

* **Documentation**
  * README and benchmarks updated to mark Better as implemented and mapped to an equivalent zstd ~7 level; Best remains pending. Added tradeoff and limitation guidance.

* **Tests**
  * New and expanded tests verify roundtrip integrity, streaming roundtrips, cross-language decoding, and size/regression expectations for Better.

### CODE REVIEW

File: cli/src/main.rs [hunk 1]
```diff
@@ -9,7 +9,7 @@ use std::path::PathBuf;
 use progress::fmt_size;
 
 use clap::{Parser, Subcommand};
-use color_eyre::eyre::{ContextCompat, WrapErr, eyre};
+use color_eyre::eyre::{eyre, ContextCompat, WrapErr};
```
[the remark this answers is not in the document]
Comment (polaz) [reply cli/src/main.rs]: @coderabbitai Fixed — ran `cargo fmt --all`.

File: zstd/src/encoding/match_generator.rs [hunk 2]
```diff
@@ -1255,6 +1358,380 @@ impl DfastMatchGenerator {
...
+            reuse_space(removed);
+        }
+    }
+
+    fn skip_matching(&mut self) {
+        self.ensure_tables();
+        let current_len = self.window.back().unwrap().len();
+        let current_abs_start = self.history_abs_start + self.window_size - current_len;
+        self.insert_positions(current_abs_start, current_abs_start + current_len);
+    }
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/match_generator.rs]: @coderabbitai Fixed — added backfill_boundary_positions() that inserts the last 3 positions of the previous slice at the start of skip_matching/start_matching, once the new slice makes them hashable.

File: zstd/src/encoding/match_generator.rs [hunk 3]
```diff
@@ -1255,6 +1359,414 @@ impl DfastMatchGenerator {
...
+        }
+        let hash = self.hash_position(&concat[idx..]);
+        // Store as (abs_pos + 1) so HC_EMPTY (0) never collides with a valid
+        // position. u32 wrapping is safe: chain_idx is masked and
+        // chain_candidates filters stale entries via history_abs_start.
+        let stored = (abs_pos as u32).wrapping_add(1);
+        let chain_idx = (stored.wrapping_sub(1) as usize) & ((1 << HC_CHAIN_LOG) - 1);
+        let prev = self.hash_table[hash];
+        self.chain_table[chain_idx] = prev;
+        self.hash_table[hash] = stored;
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/match_generator.rs]: @coderabbitai Fixed — added self-loop detection (break when next == cur) and guard at u32::MAX boundary.

File: zstd/src/encoding/match_generator.rs [hunk 4]
```diff
@@ -261,13 +324,18 @@ impl Matcher for MatchGeneratorDriver {
...
+            }
         }
 
         let mut start = 0usize;
         let mut committed_dict_budget = 0usize;
         let min_primed_tail = match self.active_backend {
             MatcherBackend::Simple => MIN_MATCH_LEN,
-            MatcherBackend::Dfast => 4,
+            MatcherBackend::Dfast | MatcherBackend::HashChain => 4,
         };
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/match_generator.rs]: @coderabbitai By design — insert_position needs 4 bytes of hash lookahead regardless of backfill. Backfill re-visits positions that already have 4+ bytes in extended history; it cannot create hashable entries from <4 byte fragments. Added clarifying code note.
